packages/coil-extension/src/background/services/Stream.ts

```diff
@@ -116,133 +111,152 @@ export class Stream extends EventEmitter {
+
+        const onMoney = (sentAmount: string) => {
+          // Wait until `setImmediate` so that `connection.totalDelivered` has been updated.
+          const receipt = stream.receipt?.toString('base64')
+          setImmediate(() => this.onMoney(connection, sentAmount, receipt))
         }
-        this._debug('error streaming. retry in 2s. err=', e.message, e.stack)
-        if (this._active) await timeout(2000)
-      } finally {
-        if (attempt) bandwidth.addSentAmount(attempt.getTotalSent())
-        if (plugin) await plugin.disconnect()
+        stream.on('outgoing_money', onMoney)
+
+        connection.on('error', async e => {
+          // Delete exhausted tokens from localstorage.
+          this._debug('stream connection error; err=%s', e.message)
+          if (!isExhaustedError(e)) return
+          if (
+            !e.ilpReject?.data.equals(
+              await sha256(Buffer.from(redeemedToken.btpToken))
+            )
+          )
+            return
+          this._anonTokens.removeToken(redeemedToken.btpToken)
+        })
+
+        connection.once('close', () => {
+          stream.removeListener('outgoing_money', onMoney)
+          // Cap paid tokens at 1.0 -- partial payment in flat-stacks may cause
+          // client-apparent overpayment (even though no overpayment occurred).
+          this._schedule.onSent(
+            Math.min(1.0, +connection.totalSent / redeemedToken.throughput / 60)
+          )
+          this._debug(
+            'connection closing; totalSent=%d sendMax=%d',
+            +connection.totalSent / redeemedToken.throughput,
+            stream.sendMax
+          )
+        })
+
+        void (async () => {
+          if (this._schedule.totalSent() === 0) {
+            await firstMinuteBandwidth(stream, redeemedToken.throughput)
+          } else {
+            const sendMax =
+              1.0 <= tokenFraction
+                ? FULL_TOKEN_AMOUNT
+                : tokenFraction * 60 * redeemedToken.throughput
+            this._debug(
+              'setSendMax tokenFraction=%d sendMax=%d',
+              tokenFraction,
+              sendMax
+            )
+            stream.setSendMax(sendMax)
+          }
+
+          let timer: NodeJS.Timer | undefined
+          // This timeout must match (or exceed) the packet timeout, otherwise the
+          // extension may try to reconnect & send packets while old packets are still
+          // in-flight farther upstream.
+          const waitTime = new Promise(
+            (_, reject) =>
+              (timer = setTimeout(() => reject(new Error('timeout')), 30e3))
+          )
+          // On error, destroy() was called internally by ilp-stream.
+          // Resolving here isn't necessary, but it prevents an extraneous destroy() call.
+          const waitFail = new Promise(resolve =>
+            connection.once('error', resolve)
+          )
+          await Promise.race([waitFail, waitTime, connection.end()])
```

Comment:
connection.end() waits for the stream.sendMax() ?

Reply:
yes